Consul.Test/SemaphoreTest.cs

```diff
@@ -363,5 +365,110 @@ public async Task Semaphore_Conflict()
+
+            var masterClient = new ConsulClient(c =>
+            {
+                c.Token = TestHelper.MasterToken;
+                c.Address = TestHelper.HttpUri;
+            });
+
+            var distributedLock2 = masterClient.CreateLock(new LockOptions(keyName)
+            {
+                SessionTTL = TimeSpan.FromSeconds(DefaultSessionTTLSeconds),
+                LockWaitTime = TimeSpan.FromSeconds(LockWaitTimeSeconds)
+            });
+
+            var distributedLock = _client.CreateLock(new LockOptions(keyName)
+            {
+                SessionTTL = TimeSpan.FromSeconds(DefaultSessionTTLSeconds),
+                LockWaitTime = TimeSpan.FromSeconds(LockWaitTimeSeconds)
+            });
+
+            var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
+
+            await distributedLock2.Acquire(); // Become "Master" with another instance first
+            var task = distributedLock.Acquire(cts.Token);
+
+            await Assert.ThrowsAsync<LockNotHeldException>(async () => await distributedLock.Acquire(cts.Token));
+
+            await distributedLock2.Release();
+            if (distributedLock.IsHeld)
+                await distributedLock2.Release();
+            await distributedLock2.Destroy();
+
+            masterClient.Dispose();
+        }
+
+        [Fact]
+        public async Task Cancelling_A_Token_When_Acquiring_A_Lock_Respects_The_Token()
+        {
+            const string key = "service/myApp/leader";
+
+            var masterInstanceClient = new ConsulClient(c =>
+            {
+                c.Token = TestHelper.MasterToken;
+                c.Address = TestHelper.HttpUri;
+            });
+
+            // Arrange
+            var distributedLock2 = masterInstanceClient.CreateLock(new LockOptions(key)
+            {
+                SessionTTL = TimeSpan.FromSeconds(DefaultSessionTTLSeconds),
+                LockWaitTime = TimeSpan.FromSeconds(LockWaitTimeSeconds)
+            });
+            var distributedLock = _client.CreateLock(new LockOptions(key)
+            {
+                SessionTTL = TimeSpan.FromSeconds(DefaultSessionTTLSeconds),
+                LockWaitTime = TimeSpan.FromSeconds(LockWaitTimeSeconds)
+            });
+            var cancellationOperationTimer = new Stopwatch();
+            var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
+
+            // Act
+            await distributedLock2.Acquire(); // Become "Master" with another instance first
+            cancellationOperationTimer.Start();
+            try
+            {
+                await distributedLock.Acquire(cts.Token);
+            }
+            catch (Exception) { }
+            cancellationOperationTimer.Stop();
+
+            // Assert
+            var stopTimeMs = cancellationOperationTimer.ElapsedMilliseconds;
+            var LockWaitTimeMs = TimeSpan.FromSeconds(15).TotalMilliseconds;
+            Assert.True(stopTimeMs < LockWaitTimeMs);
```

Comment:
I don't like this way of asserting, though it does the job. Alternatives welcome.

Reply:
I agree, the NUnit3 approach seems to be much more expressive (e.g. https://moleseyhill.com/2018-12-01-nunit-assert-that-examples.html), but I don't think that mixing XUnit and NUnit assertions would be a good idea.